whylog/teacher/__init__.py

```diff
@@ -166,18 +156,59 @@ def set_causes_relation(self, relation):
         """
         pass
 
-    def _verify(self):
-        """
-        Verifies if text patterns and constraints meet all requirements.
-        E.g it is required text pattern match its line in one way only.
-        """
-        pass
+    def _validate_primary_keys(self):
+        errors = []
+        for line_id in six.iterkeys(self._parsers):
+            primary_key = self._parsers[line_id].primary_key
+            pattern_match = self.pattern_assistant.get_pattern_match(line_id)
+            group_numbers = pattern_match.param_groups.keys()
+            if set(primary_key) - set(group_numbers):
+                errors.append(WrongPrimaryKeyProblem(primary_key, group_numbers, line_id))
+        return ValidationResult(errors=errors, warnings=[])
+
+    def _validate_pattern_names(self):
+        errors = []
+        names_blacklist = self._get_names_blacklist()
+        for line_id in six.iterkeys(self._parsers):
+            parser = self._parsers[line_id]
+            blacklist_except_name = set(names_blacklist) - set([parser.name])
+            if names_blacklist.count(parser.name) > 1 or \
+                    not self.config.is_free_parser_name(parser.name, blacklist_except_name):
+                errors.append(NotUniqueParserNameProblem(line_id))
+        return ValidationResult(errors=errors, warnings=[])
+
+    def _validate_log_type(self):
+        errors = []
+        for line_id in six.iterkeys(self._parsers):
+            parser = self._parsers[line_id]
+            if parser.log_type is None:
+                errors.append(NotSetLogTypeProblem(line_id))
+        return ValidationResult(errors=errors, warnings=[])
+
+    def _validate_constraints(self):
+        return ValidationResult.result_from_results(
+            [constraint.validate() for constraint in six.itervalues(self._constraint_base)]
+        )
 
-    def test_rule(self):
+    def validate(self):
         """
-        Simulates searching causes with already created rule.
+        Verifies if Rule is ready to save.
         """
-        pass
+
+        primary_keys_validation_result = self._validate_primary_keys()
```

Comment:
Then it should support the iteration. I suggest `ValidationResult = namedtuple(['errors', 'warnings'])`